### DISCUSSION TITLE
External extension usage

### DISCUSSION THREAD

**Discussion by adnanjpg** [CodinGame/monaco-vscode-api#169] [Q&A] [answered]
Hi there,

I'm trying to use [`TypeScript Essential Plugins`](https://marketplace.visualstudio.com/items?itemName=zardoy.ts-essential-plugins) extension in my editor but could not figure out how to. it is not specified neither in the docs nor in the examples, can anyone please give me an example of using an external extension?

**Comment by CGNonofr** [comment]
Indeed the feature is demonstrated

if you're using vite or rollup, you can follow the instruction here: https://github.com/CodinGame/monaco-vscode-api#loading-vsix-file

if not, I suggest you to use rollup to bundle your extension then digest the output in your app

**Reply by adnanjpg** [reply]
My app builds with storybook which uses webpack, so it is not possible for me to use rollup. isn't there a way to use this feature with webpack?

**Reply by CGNonofr** [reply]
There is no existing webpack plugin

Also, the rollup plugin uses virtual files which is not possible using webpack so it would probably be harder, if possible, to create that webpack plugin

Your best chance is really to have a minimal rollup config, with the vsix as entrypoint

**Reply by CGNonofr** [reply]
Here's a minimal rollup config:

```javascript
import vsixPlugin from 'vscode/rollup-vsix-plugin'

export default {
  input: 'zardoy.ts-essential-plugins-0.0.68.vsix',
  output: [{
    format: 'esm',
    dir: 'ts-essential-plugins',
    entryFileNames: 'index.js',
    assetFileNames: chunkInfo => {
      return `assets/[name][extname]`
    },
  }],
  external: ['vscode'],
  plugins: [
    vsixPlugin()
  ]
}
```
put it next to the vsix file, then run `npx rollup --config 'rollup.config.js'` and you'll get a `ts-essential-plugins` directory you can import from your code

**Reply by CGNonofr** [reply]
well you're unlucky because your extension seems to be not working, I'll have a look

**Reply by adnanjpg** [reply]
does rollup have to be the main compiler of the app or can I use this script on `postinstall` or something and still be able to use the `ts-essential-plugins` folder?

**Reply by CGNonofr** [reply]
You can just run it before running webpack by any mean you want

**Reply by CGNonofr** [reply, marked as answer]
Ok for it to work, you'll need 2 things:
- waiting for https://github.com/CodinGame/monaco-vscode-api/pull/170 to be released
- helping the vsix plugin by providing additionnal used files:

```javascript
import vsixPlugin from 'vscode/rollup-vsix-plugin'

export default {
  input: 'zardoy.ts-essential-plugins-0.0.68.vsix',
  output: [{
    format: 'esm',
    dir: 'ts-essential-plugins',
    entryFileNames: 'index.js',
    assetFileNames: chunkInfo => {
      return `assets/[name][extname]`
    },
  }],
  external: ['vscode'],
  plugins: [
    vsixPlugin({
      async getAdditionalResources () {
        return [{
          path: 'node_modules/typescript-essential-plugins/package.json'
        }, {
          path: 'node_modules/typescript-essential-plugins/index.mjs'
        }]
      }
    })
  ]
}
```

**Reply by adnanjpg** [reply]
In my monaco component code, should I import the `.vsix` file, or `index.js` from the outputs?

**Reply by CGNonofr** [reply]
The rollup build step will transform the vsix to importable javascript, then you import the generated javascript in your app

**Reply by adnanjpg** [reply]
The setting I want to enable from this extension is called `tsEssentialPlugins.fixSuggestionsSorting`. when I set it to `true` on the vsc on my machine it gives me the functionality I want. 

for my monaco component, should I import the index file  `import "../../ts-essential-plugins/index"` and then use it like:

```tsx
import { createConfiguredEditor, createModelReference } from "vscode/monaco"
import * as monaco from "monaco-editor/esm/vs/editor/editor.api"

const modelRef = await createModelReference(
  monaco.Uri.file("/tmp/test.tsx"),
  defaultEditorContents,
)
createConfiguredEditor(
  document.getElementById("editor-mountpoint")!,
  {
    model: modelRef.object.textEditorModel,
    "tsEssentialPlugins.fixSuggestionsSorting": true,
  } as monaco.editor.IStandaloneEditorConstructionOptions,
)
```

is this the way it is supposed to work?

**Reply by CGNonofr** [reply]
If you provide your configuration only to the editor, the extension will not be able to access it

instead, you can use the `updateUserConfiguration` function from `vscode/service-override/configuration`
